### PR TITLE
fix(staked-counter): mirror NFPM liquidityDelta onto currentLiquidityStaked (closes #780)

### LINKED ISSUE #780: Phantom positive stakedReserve residue: updateStakedPositionLiquidity skips currentLiquidityStaked counter update
## What to build

`updateStakedPositionLiquidity` (`src/EventHandlers/NFPM/NFPMCommonLogic.ts`) updates the Pool aggregator's `stakedTickEdges`/`stakedTickEdgeNets` and `stakedReserve0/1` when an NFP's liquidity changes via `NFPM.IncreaseLiquidity` / `NFPM.DecreaseLiquidity` events while the position is staked, but **does not** update `currentLiquidityStaked` on either the Pool or the corresponding UserStatsPerPool entity.

As a result, the next `CLGauge.Withdraw` arrives with `liquidityToStake = position.liquidity` (reflecting all in-flight Increase/Decrease deltas) while the indexer's running `currentLiquidityStaked` only ever received the original Deposit's amount. The Withdraw guard at `GaugeSharedLogic.ts:419` then underflows (`newPoolStake < 0n || newUserStake < 0n`) and `processGaugeWithdraw` **returns early**, dropping the edges/nets decrement entirely. The position's contribution stays in the staked-tick map forever — every subsequent Swap re-derives `stakedReserve0/1`/`stakedLiquidityInRange` from the polluted map, producing phantom positive residue that scales with the gap between Deposit's `liquidityToStake` and Withdraw's `liquidityToStake`.

Fix: mirror `position.liquidity` changes onto the running `currentLiquidityStaked` counter on both Pool and UserStatsPerPool so the Withdraw guard never spuriously underflows. Includes a regression test for the full Mint → Deposit → IncreaseLiquidity-while-staked → Withdraw cycle.

This is **separate from and complements** PR #773 (merged 2026-05-22). PR #773 clamps wei-scale rounding drift in `segmentStakedReserveDelta` (~hundreds of wei per pool); this issue is the systemic mechanism that was producing millions-of-dollars-of-tokens of phantom residue per pool, which #773's clamp masked rather than fixed.

## Acceptance criteria

- [ ] `updateStakedPositionLiquidity` updates `currentLiquidityStaked` on **both** Pool and UserStatsPerPool by `liquidityDelta`, in **both** code paths (normal and the `sqrtPriceX96 === 0n` early-exit)
- [ ] Regression test reproduces the bug on pre-fix code (Mint → Deposit → IncreaseLiquidity-while-staked → Withdraw cycle leaves nonzero `stakedReserve0/1`) and passes after the fix
- [ ] Local re-index over OP block range 117,044,072 → 125,000,000 produces **zero** \`withdraw exceeds current stake\` log errors (currently 1,419 with main)
- [ ] All 11 pools in `clamp_pools.json` (USDC/OP, WETH/OP, USDC/WETH, wstETH/WETH, USDC/WLD, WETH/LUSD, USDC/sDAI, USDC/mooBIFI, USDC/wstETH, WETH/DAI, USDC/sUSD) land within wei-scale of on-chain ground truth (max abs diff < 100,000 wei on values up to 10²⁴); residual sub-zero noise is then absorbed by PR #773's `NEG_STAKED_RESERVE_GUARD` clamp

## Reproduction evidence

Verified end-to-end in worktree `.claude/worktrees/debug-staked-reserve-trace` with `raw_events: true` on a trimmed OP-only config (start_block 117,044,072 → end_block 125,000,000).

Concrete example — USDC/mooBIFI gauge `0x59913758ec03a02E4b382Ec96EE45fEA9c289514`, tokenId 31357:
- Deposited block 120,425,172 with `liquidityToStake = 19,203,953,530,494`
- `NFPM.IncreaseLiquidity` block 120,425,221 log 15, delta = 141,500,788,681,522,563 → `position.liquidity` becomes 141,519,992,635,053,057
- Withdrawn block 120,623,959 with `liquidityToStake = 141,519,992,635,053,057` (matches the post-increase chain liquidity)
- Indexer's `currentLiquidityStaked` only received +19,203,953,530,494 from the Deposit → underflow on Withdraw → guard fires → +141,500,788,681,522,563 contribution stranded in `stakedTickEdges`/`stakedTickEdgeNets`

Chain-truth validation (USDC/mooBIFI, where balanced 439 Deposits ↔ 439 Withdraws and chain has 0 NFPs owned by gauge at end of indexed range, so ground truth must be 0/0):
- Pre-fix indexer: `stakedReserve0 = +36,192,912,899`, `stakedReserve1 = +1,579,096,431,386,476,403`
- Post-fix indexer: `stakedReserve0 = -516`, `stakedReserve1 = -247` (sub-wei rounding noise, absorbed by #773)

All 11 affected pools cross-validated against `NFPM.balanceOf(gauge)` + per-position `getAmountsForLiquidity` at the last stake-touching event block — max abs diff post-fix is 25,161 wei on USDC/sUSD `stakedReserve0` (values up to 2.4×10²⁴, relative error ≤ 10⁻¹⁹).

\`withdraw exceeds current stake\` log error count: 1,419 before fix → 0 after fix, across all 11 affected pools.

## Blocked by

None — can start immediately

## Related

- PR #773 (merged) — cosmetic sub-wei truncation complement
- #771 — original staked-reserve drift report
- #719 — `stakedLiquidityInRange` derivation from edges/nets

### PR DESCRIPTION
**Pull request by FonsSicca**
## Summary

Fixes #780. `updateStakedPositionLiquidity` (the function `NFPM.IncreaseLiquidity` / `DecreaseLiquidity` delegates to for staked positions) now mirrors `liquidityDelta` onto the running `currentLiquidityStaked` counter on **both** `Pool` and the staker's `UserStatsPerPool`, in **both** code paths (normal and the `sqrtPriceX96 === 0n` early-exit). The pre-fix gap let the next gauge `Withdraw` arrive with `liquidityToStake = position.liquidity` (reflecting in-flight Increase/Decrease deltas) while the indexer's counter only had the original `Deposit` amount; the guard at [`GaugeSharedLogic.ts:419`](src/EventHandlers/Gauges/GaugeSharedLogic.ts#L419) underflowed and `processGaugeWithdraw` returned early, stranding contributions in `stakedTickEdges`/`stakedTickEdgeNets` and producing phantom positive `stakedReserve0/1` residue that swap re-derived forever.

`position.owner` is the real staker because `handleRegularTransfer` skips owner updates on gauge stake/unstake transfers (see [`NFPMTransferLogic.ts:300-318`](src/EventHandlers/NFPM/NFPMTransferLogic.ts#L300-L318)), so `loadOrCreateUserData(position.owner)` resolves the same `UserStatsPerPool` that `processGaugeDeposit` originally incremented.

**Separate from and complementary to PR #773** (merged): #773 clamps wei-scale rounding drift in `segmentStakedReserveDelta` (`NEG_STAKED_RESERVE_GUARD`); this PR is the systemic mechanism behind the millions-of-dollars-of-tokens-of-phantom-residue per pool that #773's clamp was masking rather than fixing.

## Changes

- `src/EventHandlers/NFPM/NFPMCommonLogic.ts` — `updateStakedPositionLiquidity` now (a) adds `incrementalCurrentLiquidityStaked: liquidityDelta` to the pool diff in both paths and (b) loads + updates the staker's `UserStatsPerPool` in parallel with `updatePool` via `Promise.all`.
- `test/EventHandlers/NFPM/NFPMCommonLogic.test.ts` — 5 new/extended assertions covering both pool-side and user-side wiring in both code paths.
- `test/Aggregators/CLStakedLiquidityEdgeSanity.test.ts` — new test `(g) [#780]` reproduces the full **Mint → Deposit → IncreaseLiquidity-while-staked → Withdraw** cycle using the USDC/mooBIFI tokenId 31357 numbers from the issue body (`L0 = 19,203,953,530,494`, `ΔL = 141,500,788,681,522,563`) and asserts the round-trip lands `currentLiquidityStaked`, `stakedTickEdges`, `stakedTickEdgeNets`, and `stakedLiquidityInRange` all at zero. Pre-fix, `postIncrease.currentLiquidityStaked === L0 + ΔL` fails and the Withdraw guard fires.

## Acceptance criteria (#780)

- [x] `updateStakedPositionLiquidity` updates `currentLiquidityStaked` on both Pool and UserStatsPerPool by `liquidityDelta`, in both code paths — verified in `src/EventHandlers/NFPM/NFPMCommonLogic.ts`
- [x] Regression test reproduces the bug on pre-fix code and passes after the fix — see `(g) [#780]` in `test/Aggregators/CLStakedLiquidityEdgeSanity.test.ts`
- [x] Zero `withdraw exceeds current stake` log errors on OP 117,044,072 → 125,000,000 — **pre-validated** in worktree `.claude/worktrees/debug-staked-reserve-trace` (1,419 → 0 per issue body)
- [x] 11 pools in `clamp_pools.json` land within wei-scale of on-chain ground truth — **pre-validated** in same debug worktree (USDC/mooBIFI post-fix `stakedReserve0 = -516`, `stakedReserve1 = -247`; max abs diff 25,161 wei on USDC/sUSD; residue absorbed by #773's `NEG_STAKED_RESERVE_GUARD`)

## Test plan

- [x] `pnpm test` — 1430 passed / 33 skipped (no regressions)
- [x] `pnpm exec tsc --noEmit` — clean
- [x] `biome check --write` — clean (one formatting fix to the test import line)
- [ ] Optional re-validation: re-run the trimmed OP-only config in `.claude/worktrees/debug-staked-reserve-trace` (start_block 117,044,072 → 125,000,000) and confirm `withdraw exceeds current stake` count is 0 and `clamp_pools.json` chain-truth diffs stay < 100k wei

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Ensure staked liquidity deltas are mirrored into both pool and user staked-liquidity counters and updated concurrently, keeping pool/user counters aligned (including early-exit pricing path).

* **Tests**
  * Expanded regression and unit tests to cover NFPM-mediated staked liquidity changes, counter mirroring, concurrency paths, and edge-case scenarios.

[Review Change Stack](https://app.coderabbit.ai/change-stack/velodrome-finance/indexer/pull/781?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by FonsSicca** [comment]
@coderabbitai review

File: src/EventHandlers/NFPM/NFPMCommonLogic.ts [hunk 1]
```diff
@@ -90,6 +90,21 @@ export async function attributeLiquidityChangeToUserStatsPerPool(
...
+ * indexer's running counter only received the original Deposit amount; the
+ * Withdraw guard at `GaugeSharedLogic.ts:419` then underflows and the edges
+ * decrement is dropped, leaving phantom positive residue in
+ * `stakedTickEdges`/`stakedTickEdgeNets` that swap re-derives forever.
+ *
+ * Position.owner is the real staker, not the gauge: `handleRegularTransfer`
+ * skips owner updates on gauge stake/unstake transfers (see
+ * `NFPMTransferLogic.ts:300-318`), so `loadOrCreateUserData(position.owner)`
+ * resolves the UserStatsPerPool that `processGaugeDeposit` originally
+ * incremented.
```
[the remark this answers is not in the document]
Comment (FonsSicca) [reply src/EventHandlers/NFPM/NFPMCommonLogic.ts:106]: Done — added `@returns` describing the staged pool + user counter writes. Pushed in the follow-up commit.

File: test/EventHandlers/NFPM/NFPMCommonLogic.test.ts [hunk 2]
```diff
@@ -368,6 +373,52 @@ describe("NFPMCommonLogic", () => {
...
+      vi.mocked(loadOrCreateUserData).mockReset();
+      vi.mocked(updateUserStatsPerPool).mockReset();
+      const mockUserData = {
+        userAddress: stakedPosition.owner,
+        poolAddress,
+        chainId,
+        currentLiquidityStaked: 0n,
+      };
+      // biome-ignore lint/suspicious/noExplicitAny: minimal user-data shape for the mock
+      vi.mocked(loadOrCreateUserData).mockResolvedValue(mockUserData as any);
```
[the remark this answers is not in the document]
Comment (FonsSicca) [reply test/EventHandlers/NFPM/NFPMCommonLogic.test.ts]: Done — switched to `createMockUserStatsPerPool` from `setupCommon()` and dropped the `any` cast. Pushed in the follow-up commit.

**Review by WooSungD — APPROVED** [review]
